proposals/NNNN-swiftpm-additional-plugin-apis.md

```diff
@@ -0,0 +1,450 @@
+public class ExecutableProduct: Product {
+    /// The target that contains the main entry point of the executable. Every
+    /// executable product has exactly one main executable target. This target
+    /// will always be one of the targets in the product's `targets` array.
+    public let mainTarget: Target
+}
+
+/// Represents a library product defined in a package.
+public class LibraryProduct: Product {
+    /// Whether the library is static, dynamic, or automatically determined.
+    public let type: LibraryType
+}
+
+/// A type of library product.
+public enum LibraryType {
+    /// Static library.
+    case `static`
+
+    /// Dynamic library.
+    case `dynamic`
+
+    /// The type of library is unspecified and will be determined at build time.
+    case `automatic`
+}
+
+/// Represents a single target defined in a package. Specializations represent
+/// different types of targets.
+public class Target {
+    /// The name of the target, as defined in the package manifest. It's unique
+    /// among the targets of the package.
+    public var name: String
+    
+    /// The absolute path of the target directory in the local file system.
+    public var directory: Path
+    
+    /// Any other targets on which this target depends, in the same order as
+    /// they are specified in the package manifest. Conditional dependencies
+    /// that do not apply have already been filtered out.
+    public var dependencies: [Dependency]
+
+    /// Represents a dependency of a target on a product or on another target.
+    public enum Dependency {
+        /// A dependency on a target in the same package.
+        case target(Target)
+
+        /// A dependency on a product in another package.
+        case product(Product)
+    }
+}
+
+/// Represents a target consisting of a source code module, containing either
+/// Swift or source files in one of the C-based languages.
+public class SourceModuleTarget: Target {
+    /// The name of the module produced by the target (derived from the target
+    /// name, though future SwiftPM versions may allow this to be customized).
+    public var moduleName: String
+
+    /// The directory containing public C headers, if applicable. This will
+    /// only be set for targets that have Clang sources, and only if there is
+    /// a public headers directory.
+    public var publicHeadersDirectory: Path?
+
+    /// The source files that are associated with this target (any files that
+    /// have been excluded in the manifest have already been filtered out).
+    public var sourceFiles: FileList
+}
+
+/// Represents a target describing a library that is distributed as a binary.
+public class BinaryLibraryTarget: Target {
+    /// The library in the local file system.
+    public var libraryPath: Path
+}
+
+/// Represents a target describing a system library that is expected to be
+/// present on the host system.
+public class SystemLibraryTarget: Target {
```

Comment:
I think we should make the pkgConfig data available here (both the original pkgConfig name, as well as the actual flags from the pkgConfig files found, if any). Possibly it could also be interesting to retain the provider info, but I am less sure about the usefulness of that.

Reply:
Right, we have a lot of info available, but the question was what’s useful (since plugins won’t be extending existing commands through new build settings, at least not yet).  But I guess even for source generation it might be useful to have the actual pkgConfig.

Reply:
I've added this... I'll be curious to hear what you think!

It's somewhat odd to me that the pkg-config name is optional.  Is it possible to use system libraries without a package config but with just a module map and some headers?  There doesn't seem to be a way to add linker flags that don't come from a pkg-config, but I guess at least the linkage flag can be added via the module map.